src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs

```diff
@@ -36,62 +36,29 @@ private sealed class StateSet
 
-                var (_, version) = _analyzer.GetAnalyzerIdAndVersion();
-                _analyzerVersion = version;
+                var (_, version) = Analyzer.GetAnalyzerIdAndVersion();
+                AnalyzerVersion = version;
 
-                _analyzerStateNames = AnalyzerStateNames.For(_analyzer);
+                _persistantNames = PersistentNames.Create(Analyzer);
 
                 _activeFileStates = new ConcurrentDictionary<DocumentId, ActiveFileState>(concurrencyLevel: 2, capacity: 10);
                 _projectStates = new ConcurrentDictionary<ProjectId, ProjectState>(concurrencyLevel: 2, capacity: 1);
             }
 
-            public string StateName => _analyzerStateNames.StateName;
-            public string SyntaxStateName => _analyzerStateNames.SyntaxStateName;
-            public string SemanticStateName => _analyzerStateNames.SemanticStateName;
-            public string NonLocalStateName => _analyzerStateNames.NonLocalStateName;
-
-            public string Language => _language;
-            public string ErrorSourceName => _errorSourceName;
-
-            public DiagnosticAnalyzer Analyzer => _analyzer;
-            public VersionStamp AnalyzerVersion => _analyzerVersion;
+            public string StateName => _persistantNames.StateName;
+            public string SyntaxStateName => _persistantNames.SyntaxStateName;
+            public string SemanticStateName => _persistantNames.SemanticStateName;
+            public string NonLocalStateName => _persistantNames.NonLocalStateName;
 
             [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/34761", AllowCaptures = false, AllowGenericEnumeration = false)]
             public bool ContainsAnyDocumentOrProjectDiagnostics(ProjectId projectId)
             {
-                if (ContainsAnyDocumentDiagnostics(_activeFileStates, projectId))
-                {
-                    return true;
-                }
-
-                if (!_projectStates.TryGetValue(projectId, out var projectState))
-                {
-                    return false;
-                }
-
-                return !projectState.IsEmpty();
-
-                static bool ContainsAnyDocumentDiagnostics(ConcurrentDictionary<DocumentId, ActiveFileState> activeFileStates, ProjectId projectId)
-                {
-                    foreach (var kvp in activeFileStates)
-                    {
-                        var documentId = kvp.Key;
-                        if (documentId.ProjectId == projectId)
-                        {
-                            var state = kvp.Value;
-                            if (!state.IsEmpty)
-                            {
-                                return true;
-                            }
-                        }
-                    }
-
-                    return false;
-                }
+                return _activeFileStates.Values.Any(state => state.DocumentId.ProjectId == projectId && !state.IsEmpty) ||
```

Comment:
This creates captures and the method is annotated with `AllowCaptures=false`

Reply:
Good catch, I totally missed that attribute :-o. Looks like we don't have analyzers validating this :(

Reply:
The attribute says `AllowGenericEnumeration = false` but even the code before my change enumerates `ConcurrentDictionary`, which allocates an enumerator.

Reply:
#resolved